New module: Add kinit (identity/kinit)

bincyber opened the pull request:
##### SUMMARY

Added a new module for obtaining a short-lived Kerberos ticket-granting ticket (TGT).

Some cli tools such as `ipa` use Kerberos for authentication thus they require obtaining a valid Kerberos ticket before they are executed.

Currently, this can be accomplished using the expect module:
```
- name: expect | acquire a short-lived Kerberos ticket
  expect:
    command: "/usr/bin/kinit {{ kerberos_principal }} -l 60"
    responses:
      "Password for.*": "{{ password }}"
  no_log: true

- name: shell | register an HTTP service in FreeIPA for this node
  shell: "/usr/bin/ipa service-add HTTP/{{ ansible_fqdn }} > /etc/ipa/service-add-http.log"
  args:
    creates: "/etc/ipa/service-add-http.log"
```
or the shell module:
```
- name: shell | acquire a short-lived Kerberos ticket
  shell: "echo {{ password }} | /usr/bin/kinit {{ kerberos_principal }} -l 60"
  no_log: true
```

However, the expect module requires a newer version of pexpect installed on the host (which is only available via pip on RHEL/CentOS machines and not yum). Also, since `no_log` is used, it hides error messages if the task fails and does not log. It is possible to use environment variables with the shell module to avoid the need to use `no_log`:
```
- name: shell | acquire a short-lived Kerberos ticket
  shell: "echo $KERBEROS_PASSWORD | /usr/bin/kinit {{ kerberos_principal }} -l 60"
  environment:
    KERBEROS_PASSWORD: "{{ password }}"
```
But I'd prefer a native module.

##### ISSUE TYPE
 - New Module Pull Request


##### COMPONENT NAME

kinit

##### ANSIBLE VERSION

```
ansible 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'modules/']
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```

##### ADDITIONAL INFORMATION

bincyber committed:
Added new module: kinit

- added a new module which is a wrapper around kinit to obtain
  a short-lived Kerberos ticket-granting ticket (TGT)

alikins commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,116 @@
+version_added: "2.4"
+author: "Ali (@bincyber)"
+short_description: Obtain a Kerberos ticket-granting ticket
+description:
```
I'd like to see the documentation describe some of the security tradeoffs involved in this approach  (ie, sending krb passwords over the wire)

bincyber replied:
I'm not aware of any security tradeoffs in regards to how the password is supplied to the module. It's exactly the same as how passwords are supplied to other modules that support a password parameter: `no_log` is used to prevent the password from being logged or displayed in verbose mode.

However, I will update the documentation to make a note of why using this module is preferably to expect/shell from a security perspective.

bincyber committed:
Updated documentation for kinit module

bincyber commented:
ready_for_review

bincyber commented:
What do I need to do to get this module merged?

alxgu commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,122 @@
+DOCUMENTATION = '''
+---
+module: kinit
+version_added: "2.4"
```
2.4 is alreay released. Use 2.5

alxgu commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,122 @@
+    password = module.params['password']
+    lifetime = module.params.get('lifetime', 60)
+
+    kinit = find_executable('kinit')
```
the AnsibleModule class has a method to find executables. Use that method:
kinit_cmd = module.get_bin_path("kinit", required=True)

bincyber replied:
Done. Thanks

bincyber committed:
Updated kinit module

bincyber commented:
Label should be updated to `affects_2.5`

jhawkesworth commented:
I like the idea of this, but I'd like it even more if it was a bit more 'moduely' i.e it didn't always just acquire a new ticket, but instead ran a klist and checked if the remaining validity period was >= lifetime.

Probably overkill really but  I guess if it didn't allways run kinit you could use it in playbooks that overlap at runtime without a reduced risk of the ticket being temporarily invalidated by another playbook at runtime.  I think that would make it more generally applicable.

Having check mode would be nice too.  I appreciate the stuff above would mean having to parse the klist output, which would add to the complexity of the module, but adding idempotence and check mode would make it fit even better with the rest of ansible.

fxfitz commented:
+1 for what @jhawkesworth suggested, but wouldn't necessarily need to hold up the module for it.

bincyber committed:
Modified kinit module

- added idempotency
- added support for check mode
- added new 'state' parameter to allow generating or deleting Kerberos tickets

bincyber commented:
@jhawkesworth I support that idea and will be pushing a new commit to include those requirements.

bincyber commented:
Triggering CI. Please review this PR.

puiterwijk commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,203 @@
+    default: null
+  lifetime:
+    description:
+      - The lifetime of the Kerberos ticket.
```
Maybe just add a "in seconds" here?

puiterwijk commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,203 @@
+        return datetime.datetime.fromtimestamp(time.time() + self.lifetime)
+
+    def get_existing_tickets(self):
+        return self.module.run_command([self.klist])
```
Do note that this only looks at the primary credential cache. This means that if you use a collection credential cache such as DIR, KEYRING or KCM (default in most distributions these days), you might not see a ticket that exists in a non-primary credential cache.
Maybe consider using `klist -A`, or try a `kswitch -p <principal>`?

The latter is easier to parse, but might break stuff if people are still using the old krb5 APIs (rather than GSSAPI) and were using the primary credential that you just made secondary.

puiterwijk commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,203 @@
+            self.module.exit_json(changed=True, msg='Obtained a short-lived Kerberos ticket', rc=rc)
+
+    def destroy_existing_tickets(self):
+        rc, out, err = self.module.run_command([self.kdestroy, '-q'])
```
Again, this only works on the primary credential cache. If the ticket is in another, this might not work.

puiterwijk commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,203 @@
+
+    kerberos_ticket = KerberosTicket(module)
+
+    rc, out, err = kerberos_ticket.get_existing_tickets()
```
Do note that this does not take `self.principal` into account, neither is that being checked. So you might return with `changed=False` if there's a tgt for a totally different principal.

puiterwijk reviewed (changes requested):
Looks like a nice module that I could definitely use, thanks!
I did have some comments about use for credential cache collections where this module would possibly do the wrong thing.
Also, do note that `get_existing_tickets` does not use `self.principal` to check that the current ticket is issued for the correct principal.
I personally think that the principal check really needs to be fixed (as it could cause a lot of confusion), the collection notes would be ideal to get fixed, but as long as the principal is checked that should be less of a problem.

bincyber commented:
@puiterwijk Thanks for providing feedback on this PR. I will modify the module to check all collection credential caches using `kinit -A` and check that any existing TGT is for the right principal.

bincyber committed:
Updated kinit module

- module now checks all collection credential caches for a TGT
  for the specified principal

bincyber committed:
Fixed issues reported by sanity tests

puiterwijk commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,256 @@
+                module.exit_json(changed=True, msg='Obtained a Kerberos ticket', cache_name=cache_name)
+
+        elif kerberos_ticket.state == 'absent':
+            if module.check_mode:
```
Why do you have a different code path here?
The non-check line does the same but provides a bit more info, so just using that in the check case should be fine I'd say.

puiterwijk commented on lib/ansible/modules/identity/kinit.py:
```diff
@@ -0,0 +1,256 @@
+                else:
+                    module.exit_json(changed=False)
+
+            rc, err = kerberos_ticket.renew_ticket()
```
I'd say that if there is still a ticket that has a validity of more than was requested, we don't need to try to renew and can just return a changed=False, msg='We already had a sufficient ticket'?

bincyber replied:
That's true. I will update the module.

puiterwijk reviewed (approved):
Looks fine to me, just a few suggestions.

bincyber committed:
Updated kinit module with minor fixes

bincyber commented:
ready_for_review

bincyber commented:
It's been over a year since I've opened this PR. Can we get this merged please?

puiterwijk reviewed (approved):
While the nested if's in the `main` function seems slightly clunky, I think this code looks reasonable from a krb5 point of view.

maxamillion commented:
@puiterwijk thank you

maxamillion commented:
@alikins I'm good with it if you are, but if there's more to discuss here then let's add it to the community meeting agenda.

bincyber commented:
Thanks for reviewing @puiterwijk. @gundalow can this be merged please?

gundalow committed:
version_added to 2.8. Name tasks

gundalow commented:
@bincyber Hi, I've pushed a change to update a few bits of docs.

I've asked @alikins to do a final review

lhoss commented:
Great to soon have a kinit module 👍 
I am only missing the option to kinit based on a *keytab* (instead of a password)

In the meantime I am using my new ansible role, published today, which has a nice lookup feature (useful for people running kerberized Cloudera clusters):
https://github.com/lhoss/kerberos_kinit
